core/src/main/java/brooklyn/catalog/internal/CatalogInitialization.java

```diff
@@ -0,0 +1,377 @@
+
+    /*
+
+    A1) if not persisting, go to B1
+    A2) if --catalog-reset, delete persisted catalog items
+    A3) read persisted catalog items (possibly deleted in A2), go to C1
+    A4) go to B1
+
+    B1) look for --catalog-initial, if so read it, then go to C1
+    B2) look for BrooklynServerConfig.BROOKLYN_CATALOG_URL, if so, read it, supporting YAML or XML (warning if XML), then go to C1
+    B3) look for ~/.brooklyn/catalog.bom, if exists, read it then go to C1
+    B4) look for ~/.brooklyn/brooklyn.xml, if exists, warn, read it then go to C1
+    B5) read all classpath://brooklyn/default.catalog.bom items, if they exist (and for now they will)
+    B6) go to C1
+
+    C1) if --catalog-add, read and add those items
+
+    D1) if persisting, read the rest of the persisted items (entities etc)
+
+     */
+
+    private static final Logger log = LoggerFactory.getLogger(CatalogInitialization.class);
+    
+    String initialUri;
+    boolean reset;
+    String additionsUri;
+    boolean force;
+
+    boolean disallowLocal = false;
+    List<Function<CatalogInitialization, Void>> callbacks = MutableList.of();
+    boolean hasRunBestEffort = false, hasRunOfficial = false, isPopulating = false;
+    
+    ManagementContext managementContext;
+    boolean isStartingUp = false;
+    boolean failOnStartupErrors = false;
+    
+    Object mutex = new Object();
+    
+    public CatalogInitialization(String initialUri, boolean reset, String additionUri, boolean force) {
+        this.initialUri = initialUri;
+        this.reset = reset;
+        this.additionsUri = additionUri;
+        this.force = force;
+    }
+    
+    public CatalogInitialization() {
+        this(null, false, null, false);
+    }
+
+    public void injectManagementContext(ManagementContext managementContext) {
+        if (this.managementContext!=null && managementContext!=null && !this.managementContext.equals(managementContext))
+            throw new IllegalStateException("Cannot switch management context of "+this+"; from "+this.managementContext+" to "+managementContext);
+        this.managementContext = managementContext;
+    }
+    
+    public ManagementContext getManagementContext() {
+        return Preconditions.checkNotNull(managementContext, "management context has not been injected into "+this);
+    }
+
+    public CatalogInitialization addPopulationCallback(Function<CatalogInitialization, Void> callback) {
+        callbacks.add(callback);
+        return this;
+    }
+
+    public boolean isInitialResetRequested() {
+        return reset;
+    }
+
+    public boolean hasRunOfficial() { return hasRunOfficial; }
+    public boolean hasRunIncludingBestEffort() { return hasRunOfficial || hasRunBestEffort; }
+
+    /** makes or updates the mgmt catalog, based on the settings in this class */
+    public void populateCatalog(boolean needsInitial, Collection<CatalogItem<?, ?>> optionalItemsForResettingCatalog) {
+        try {
+            isPopulating = true;
```

Comment:
To @robertgmoss's point, why deliberately do the `isPopulating` and the finally block outside of the mutex?

For example, the other method to be synchronized (`populateBestEffort `) grabs the mutex and the does:
    `if (hasRunOfficial || hasRunBestEffort || isPopulating) return;`
Therefore `populateBestEffort ` return immediately even though this method hasn't actually populated the catalog yet.